drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1020 @@
+	struct can_nxp_s32_data *data = dev->data;
+	int err;
+
+	if (data->started) {
+		return -EALREADY;
+	}
+
+	if (config->phy != NULL) {
+		err = can_transceiver_enable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to enable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	data->started = true;
+
+	return 0;
+}
+
+static int can_nxp_s32_abort_msg(const struct can_nxp_s32_config *config, int mb_idx)
+{
+	uint32_t time_start = 0;
+	int ret = 0;
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	CanXL_ClearMsgBuffIntCmd(config->base_grp_ctrl, mb_idx);
+	CanXL_ClearMsgDescIntStatusFlag(config->base_grp_ctrl, mb_idx);
+
+	time_start = k_uptime_get();
+	/* Set system lock Status */
+	(void)config->base_dsc_ctrl->DSCMBCTRLAR[mb_idx].SYSLOCK.DCSYSLOCK;
+	while (CanXL_GetDescControlStatus(config->base_dsc_ctrl, mb_idx)
+			== CANEXCEL_DESCNTSTATUS_LOCKED_HW) {
+		if (k_uptime_get() - time_start >= CAN_NXP_S32_TIMEOUT_MS) {
+			ret = CANEXCEL_STATUS_TIMEOUT;
+			break;
+		}
+	}
+
+	/* Inactive descriptor */
+	config->base_dsc_ctrl->DSCMBCTRLAR[mb_idx].ACT.DCACT = 0;
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	return ret;
+}
+
+static int can_nxp_s32_stop(const struct device *dev)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	can_tx_callback_t function;
+	void *arg;
+	int alloc;
+	int err;
+
+	if (!data->started) {
+		return -EALREADY;
+	}
+
+	data->started = false;
+
+	/* Abort any pending TX frames before entering freeze mode */
+	for (alloc = 0; alloc < CAN_NXP_S32_MAX_TX; alloc++) {
+		function = data->tx_cbs[alloc].function;
+		arg = data->tx_cbs[alloc].arg;
+
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			if (can_nxp_s32_abort_msg(config,
+					ALLOC_IDX_TO_TXMB_IDX(alloc))) {
+				LOG_ERR("Can't abort message !");
+			};
+
+			function(dev, -ENETDOWN, arg);
+			k_sem_give(&data->tx_allocs_sem);
+		}
+	}
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+
+static int can_nxp_s32_set_mode(const struct device *dev, can_mode_t mode)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_ModesType can_nxp_s32_mode = CAN_MODE_NORMAL;
+	bool canfd = false;
+	bool brs = false;
+
+	if (data->started) {
+		return -EBUSY;
+	}
+#ifdef CONFIG_CAN_FD_MODE
+	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_FD)) != 0) {
+#else
+	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY)) != 0) {
+#endif
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	if ((mode & (CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY))
+				== (CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY)) {
+		LOG_ERR("unsupported both modes: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	canfd = !!(mode & CAN_MODE_FD);
+	brs = canfd ? true : false;
```

Comment:
Is this `brs` variable just for enabling BRS when receiving? TX frame BRS most not be enabled (according to the current Zephyr CAN API) just based on `CAN_MODE_FD`.

Since `canfd` is already a boolean, you could just assign its value directly to `brs`.